lib/screenplay/suppressed_predictions/suppressed_prediction.ex

```diff
@@ -0,0 +1,172 @@
+
+  @primary_key false
+  schema "suppressed_predictions" do
+    field(:location_id, :string, primary_key: true)
+    field(:route_id, :string, primary_key: true)
+    field(:direction_id, :integer, primary_key: true)
+    field(:clear_at_end_of_day, :boolean)
+
+    timestamps(type: :utc_datetime)
+  end
+
+  def changeset(suppressed_prediction, attrs \\ %{}) do
+    places = Places.get()
+
+    suppressed_prediction
+    |> cast(attrs, [
+      :location_id,
+      :route_id,
+      :direction_id,
+      :clear_at_end_of_day
+    ])
+    |> validate_required([
+      :location_id,
+      :route_id,
+      :direction_id,
+      :clear_at_end_of_day
+    ])
+    |> validate_number(:direction_id,
+      less_than_or_equal_to: 1,
+      greater_than_or_equal_to: 0,
+      message: "Location Direction ID must be 0 or 1"
+    )
+    |> validate_edge_cases()
+    |> validate_location(places)
+    |> validate_routes_with_location_id(places)
+  end
+
+  defp validate_edge_cases(changeset) do
+    location_id = get_field(changeset, :location_id)
+    route_id = get_field(changeset, :route_id)
+
+    cond do
+      # JFK/UMass has the Ashmont/Braintree split, we ask for those stop ids rather than the parent id
+      location_id == "place-jfk" ->
+        add_error(
+          changeset,
+          :location_id,
+          "Please provide either the Ashmont or Braintree platform location_id: (#{@jfk_umass_ashmont_place} or #{@jfk_umass_braintree_place})"
+        )
+
+      # We don't handle separate Green Line and Silver Line branching route_ids
+      # That is populated when we pass data to Transit Data
+      # For internal use we just use "Green" or "Silver" to show all green/silver line routes
+      route_id in @green_line_routes ->
+        add_error(
+          changeset,
+          :route_id,
+          "Please provide just `Green` as the route_id for handling all Green Line routes"
+        )
+
+      true ->
+        changeset
+    end
+  end
+
+  defp validate_location(changeset, places) do
+    location_id = get_field(changeset, :location_id)
+
+    if places |> Enum.any?(&(&1.id == location_id)) do
+      changeset
+    else
+      add_error(changeset, :location_id, "Location `#{location_id}` does not exist")
+    end
+  end
+
+  defp validate_routes_with_location_id(changeset, places) do
+    location_id = get_field(changeset, :location_id)
+
+    if location_id in [@jfk_umass_ashmont_place, @jfk_umass_braintree_place] do
+      "place-jfk"
+    else
+      location_id
+    end
+    |> validate_routes(changeset, places)
+  end
+
+  defp validate_routes(location_id, changeset, places) do
+    route_id = get_field(changeset, :route_id)
+
+    places
+    |> Enum.find(fn place -> place.id == location_id end)
+    |> case do
+      nil ->
+        add_error(
+          changeset,
+          :route_id,
+          "Location `#{location_id}` does not exist"
+        )
+
+      place ->
+        if route_id in ["Green" | place.routes] do
```

Comment:
I had to read this logic block a couple times to convince myself it was correct. The front-end code has a similar bit of logic, and I found it helpful to branch first on `route_id`, to split out the Silver and Green cases, and go from there. FYI, we don't need to (or want to, really) support specifying branches, e.g. "Green-D", since that won't be an option in the UI. Also, since we'll design the UI to avoid any validation errors, we can afford to be less specific and just roll things up into a generic error message, which should simplify things here.

Reply:
Sounds good!

Assuming the front end UI can clear up some of the route_id handling, I basically was able to strip out the logic to be only the Silver Line handling (Green Line will just pass through and the other routes will pass through.)

I'm hoping I understood what you're saying correctly, but let me know if there's something I missed with my new simplified validation.